relayer/cmd/fetch_messages.go

```diff
@@ -104,75 +120,99 @@ func getEthContractEventsAndTrie(
 	}
 	trie := makeTrie(receipts)
 
-	contracts, err := getEthContractsFromConfig(config, conn.GetClient())
+	allEvents := make([]*gethTypes.Log, 0)
+
+	basicEvents, err := getEthBasicMessages(ctx, basicOutboundChannel, block.NumberU64(), index)
 	if err != nil {
 		return nil, nil, err
 	}
+	allEvents = append(allEvents, basicEvents...)
 
-	contractMessages, err := getEthContractMessages(ctx, contracts, block.NumberU64(), index)
+	incentivizedEvents, err := getEthIncentivizedMessages(ctx, incentivizedOutboundChannel, block.NumberU64(), index)
 	if err != nil {
 		return nil, nil, err
 	}
+	allEvents = append(allEvents, incentivizedEvents...)
 
-	return contractMessages, trie, nil
+	return allEvents, trie, nil
 }
 
-func getEthContractsFromConfig(config *ethereum.Config, client *ethclient.Client) ([]*outbound.Contract, error) {
-	contractBasic, err := outbound.NewContract(gethCommon.HexToAddress(config.Channels.Basic.Outbound), client)
+func getEthBasicMessages(
+	ctx context.Context,
+	contract *outbound.BasicOutboundChannel,
+	blockNumber uint64,
+	index uint64,
+) ([]*gethTypes.Log, error) {
+	events := make([]*gethTypes.Log, 0)
+	filterOps := bind.FilterOpts{Start: blockNumber, End: &blockNumber, Context: ctx}
+
+	iter, err := contract.FilterMessage(&filterOps)
 	if err != nil {
 		return nil, err
 	}
 
-	contractIncentivized, err := outbound.NewContract(gethCommon.HexToAddress(config.Channels.Incentivized.Outbound), client)
-	if err != nil {
-		return nil, err
+	for {
+		more := iter.Next()
+		if !more {
+			err = iter.Error()
+			if err != nil {
+				return nil, err
+			}
+			break
+		}
+
+		if uint64(iter.Event.Raw.TxIndex) != index {
+			continue
+		}
+		events = append(events, &iter.Event.Raw)
 	}
 
-	return []*outbound.Contract{contractBasic, contractIncentivized}, nil
+	return events, nil
 }
 
-func getEthContractMessages(
+func getEthIncentivizedMessages(
 	ctx context.Context,
-	contracts []*outbound.Contract,
+	contract *outbound.IncentivizedOutboundChannel,
 	blockNumber uint64,
 	index uint64,
-) ([]*outbound.ContractMessage, error) {
-	events := make([]*outbound.ContractMessage, 0)
+) ([]*gethTypes.Log, error) {
+	events := make([]*gethTypes.Log, 0)
 	filterOps := bind.FilterOpts{Start: blockNumber, End: &blockNumber, Context: ctx}
-	for _, contract := range contracts {
-		iter, err := contract.FilterMessage(&filterOps)
-		if err != nil {
-			return nil, err
-		}
 
-		for {
-			more := iter.Next()
-			if !more {
-				err = iter.Error()
-				if err != nil {
-					return nil, err
-				}
-				break
-			}
+	iter, err := contract.FilterMessage(&filterOps)
+	if err != nil {
+		return nil, err
+	}
 
-			if uint64(iter.Event.Raw.TxIndex) != index {
-				continue
+	for {
+		more := iter.Next()
+		if !more {
+			err = iter.Error()
+			if err != nil {
+				return nil, err
 			}
-			events = append(events, iter.Event)
```

Comment:
Was this a bug? (appending an event when there were none even if there was no error). As above, I ask to see if I'm getting the semantics of the function right.
BTW I think the code would be clearer if directly returns `return events, nil` instead of breaking.

Reply:
I can't see a bug here. We were never calling `append` within the `if !more` block. The patch diff is just very messy which can be misleading.

Agree on the `return events, nil` instead of breaking, however my VS code linter complains about it, so leaving it in for now.

Reply:
The diff was misleading: I read it as this line `events = append(events, iter.Event)` right after 
```go
			err = iter.Error()
			if err != nil {
				return nil, err
			}
                        // here
```
which wasn't the case looking at line numbers.